src/compiler.ts

```diff
@@ -4617,79 +4619,207 @@ export class Compiler extends DiagnosticEmitter {
+                break;
+              }
+              assert(prototype.kind == ElementKind.FUNCTION_PROTOTYPE);
+              this.i32PowInstance = instance = this.resolver.resolveFunction(<FunctionPrototype>prototype, null);
+            }
+            if (!instance || !this.compileFunction(instance)) {
               expr = module.unreachable();
-              break;
+            } else {
+              expr = this.makeCallDirect(instance, [ leftExpr, rightExpr ], expression);
             }
-            let prototype = namespace.members ? namespace.members.get(CommonNames.pow) : null;
-            if (!prototype) {
-              this.error(
-                DiagnosticCode.Cannot_find_name_0,
-                expression.range, "Mathf.pow"
-              );
+            break;
+          }
+          case TypeKind.I64:
+          case TypeKind.U64: {
+            instance  = this.i64PowInstance;
+            if (!instance) {
+              let prototype = this.program.lookupGlobal(CommonNames.ipow64);
+              if (!prototype) {
+                this.error(
+                  DiagnosticCode.Cannot_find_name_0,
+                  expression.range, "ipow64"
+                );
+                expr = module.unreachable();
+                break;
+              }
+              assert(prototype.kind == ElementKind.FUNCTION_PROTOTYPE);
+              this.i64PowInstance = instance = this.resolver.resolveFunction(<FunctionPrototype>prototype, null);
+            }
+            if (!instance || !this.compileFunction(instance)) {
               expr = module.unreachable();
-              break;
+            } else {
+              expr = this.makeCallDirect(instance, [ leftExpr, rightExpr ], expression);
             }
-            assert(prototype.kind == ElementKind.FUNCTION_PROTOTYPE);
-            this.f32PowInstance = instance = this.resolver.resolveFunction(<FunctionPrototype>prototype, null);
+            break;
           }
-
-        // Math.pow otherwise (result is f64)
-        // TODO: should the result be converted back?
-        } else {
-          leftExpr = this.convertExpression(leftExpr,
-            this.currentType, Type.f64,
-            false, false,
-            left
-          );
-          leftType = this.currentType;
-          rightExpr = this.compileExpression(right, Type.f64, Constraints.CONV_IMPLICIT);
-          rightType = this.currentType;
-          instance = this.f64PowInstance;
-          if (!instance) {
-            let namespace = this.program.lookupGlobal(CommonNames.Math);
-            if (!namespace) {
-              this.error(
-                DiagnosticCode.Cannot_find_name_0,
-                expression.range, "Math"
+          case TypeKind.ISIZE:
+          case TypeKind.USIZE: {
+            let type = this.options.isWasm64
+              ? (this.currentType.is(TypeFlags.SIGNED) ? Type.i64 : Type.u64)
+              : (this.currentType.is(TypeFlags.SIGNED) ? Type.i32 : Type.u32);
```

Comment:
This looks as if it'd change the type from `isize` to `i64` for example, which seems odd. Should just stay `isize`, so if the input is `isize`, the output is as well.

Reply:
Good catch. I simplified this